sw/device/tests/dif/dif_otbn_smoketest.c

```diff
@@ -2,210 +2,33 @@
-   */
-  const char *dmem_end;
-} otbn_app_t;
-
-/**
- * A function (entry point) in an OTBN application.
- */
-typedef struct otbn_func {
-  /**
-   * Application associated with this function.
-   */
-  const otbn_app_t *app;
-  /**
-   * Address of the function entry point in normal CPU address space.
-   */
-  const char *func;
-} otbn_func_t;
+OTBN_DECLARE_APP_SYMBOLS(barrett384);
+OTBN_DECLARE_PTR_SYMBOL(barrett384, wrap_barrett384);
 
-// Prepare info structs representing the barrett384 and err_test applications,
-// and the wrap_barrett384() and wrap_err_test functions. The goal of these
-// structs is to make the symbols for the application image available in a form
-// that is easier to use.
-//
-// The symbols are generated by the linker script (the start/end symbols for
-// instruction and data memories), as well as by the assembly sources (all
-// global symbols, see `sw/otbn/code-snippets/barrett384.s`).
-// Symbols are then prefixed with "_otbn_app_APPNAME_" by objcopy, as called
-// from `util/otbn_build.py`.
-//
-// TODO: Think about macro-ifying or auto-generating some of this code in a
-// future OTBN driver.
-extern const char _otbn_app_barrett384__imem_start[];
-extern const char _otbn_app_barrett384__imem_end[];
-extern const char _otbn_app_barrett384__dmem_start[];
-extern const char _otbn_app_barrett384__dmem_end[];
-extern const char _otbn_app_barrett384_wrap_barrett384[];
+static const otbn_app_t kOtbnAppBarrett384 = OTBN_APP_T_INIT(barrett384);
+static const otbn_ptr_t kOtbnFuncBarrett384WrapBarrett384 =
+    OTBN_PTR_T_INIT(barrett384, wrap_barrett384);
 
-static const otbn_app_t kOtbnAppBarrett384 = {
-    .imem_start = _otbn_app_barrett384__imem_start,
-    .imem_end = _otbn_app_barrett384__imem_end,
-    .dmem_start = _otbn_app_barrett384__dmem_start,
-    .dmem_end = _otbn_app_barrett384__dmem_end,
-};
+OTBN_DECLARE_APP_SYMBOLS(err_test);
+OTBN_DECLARE_PTR_SYMBOL(err_test, err_test);
 
-static const otbn_func_t kOtbnFuncBarrett384WrapBarrett384 = {
-    .app = &kOtbnAppBarrett384,
-    .func = _otbn_app_barrett384_wrap_barrett384,
-};
-
-extern const char _otbn_app_err_test__imem_start[];
-extern const char _otbn_app_err_test__imem_end[];
-extern const char _otbn_app_err_test__dmem_start[];
-extern const char _otbn_app_err_test__dmem_end[];
-extern const char _otbn_app_err_test_err_test[];
-
-static const otbn_app_t kOtbnAppErrTest = {
-    .imem_start = _otbn_app_err_test__imem_start,
-    .imem_end = _otbn_app_err_test__imem_end,
-    .dmem_start = _otbn_app_err_test__dmem_start,
-    .dmem_end = _otbn_app_err_test__dmem_end,
-};
-
-static const otbn_func_t kOtbnFuncErrTestWrapErrTest = {
-    .app = &kOtbnAppErrTest,
-    .func = _otbn_app_err_test_err_test,
-};
+static const otbn_app_t kOtbnAppErrTest = OTBN_APP_T_INIT(err_test);
+static const otbn_ptr_t kOtbnFuncErrTestWrapErrTest =
+    OTBN_PTR_T_INIT(err_test, err_test);
 
 const test_config_t kTestConfig = {
     .can_clobber_uart = false,
```

Comment:
This is the implicit default. You can just write `= {};`. Perhaps worth moving closer to the main function.